apps/cnspec/cmd/backgroundjob/checkin.go

```diff
@@ -0,0 +1,154 @@
+// Copyright (c) Mondoo, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package backgroundjob
+
+import (
+	"context"
+	"math/rand"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog/log"
+	"github.com/spf13/viper"
+	"go.mondoo.com/cnquery/v9"
+	"go.mondoo.com/cnquery/v9/cli/sysinfo"
+	"go.mondoo.com/cnquery/v9/providers-sdk/v1/upstream"
+	"go.mondoo.com/ranger-rpc"
+	"go.mondoo.com/ranger-rpc/plugins/scope"
+)
+
+type checkinPinger struct {
+	ctx        context.Context
+	interval   time.Duration
+	quit       chan struct{}
+	wg         sync.WaitGroup
+	endpoint   string
+	httpClient *http.Client
+	mrn        string
+	creds      *upstream.ServiceAccountCredentials
+}
+
+func NewCheckinPinger(ctx context.Context, httpClient *http.Client, endpoint string, config *upstream.UpstreamConfig, interval time.Duration) *checkinPinger {
+	return &checkinPinger{
+		ctx:        ctx,
+		interval:   interval,
+		quit:       make(chan struct{}),
+		endpoint:   endpoint,
+		httpClient: httpClient,
+		mrn:        config.Creds.Mrn,
+		creds:      config.Creds,
+	}
+}
+
+func (c *checkinPinger) Start() {
+	// determine information about the client
+	sysInfo, err := sysinfo.GatherSystemInfo()
```

Comment:
Yes, this uses the os provider. And yes it gets killed. I checked with `ps`. As soon as the actual scan finishes, all started plugins are gone.

Reply:
I think that's because there is a coordinator shutdown invoked after a scan is done. However, the scans run every ~1h and this is every ~2h + potential jitter. I guess that means we can have 1 OS provider running for a while until a scan happens

Reply:
We also have the provider heartbeat implemented though so I guess that fine. Just wanted to point out the behaviour here